File: libraries/rust/src/nats.rs
use std::collections::BTreeMap;

use serde::Serialize;
use thiserror::Error;

use crate::{
    identity::{DeckrMessage, MessageTarget},
    keys::encode_key_token,
};

const LANE_PREFIX: &str = "deckr.lane";

#[derive(Debug, Error)]
pub enum NatsBindingError {
    #[error("NATS subject disagrees with Deckr envelope sender")]
    SubjectMismatch,
    #[error("NATS header {0} disagrees with Deckr envelope")]
    HeaderMismatch(String),
}

pub fn subject_for(message: &DeckrMessage) -> String {
    format!(
        "{LANE_PREFIX}.{}.{}.{}",
        encode_key_token(&message.lane),
        encode_key_token(&message.sender.family),
        encode_key_token(&message.sender.endpoint_id)
    )
}

pub fn recipient_header(message: &DeckrMessage) -> String {
    match &message.recipient {
        MessageTarget::Endpoint(target) => target.endpoint.to_string(),
        MessageTarget::Broadcast(target) => {
            format!("broadcast:{}:{}", target.scope, target.endpoint_family)
        }
    }
}

pub fn headers_for(message: &DeckrMessage) -> BTreeMap<String, String> {
    let mut headers = BTreeMap::new();
    headers.insert("Deckr-Message-Id".to_string(), message.message_id.clone());
    headers.insert(
        "Deckr-Message-Type".to_string(),
        message.message_type.clone(),
    );
    headers.insert("Deckr-Sender".to_string(), message.sender.to_string());
    headers.insert(
        "Deckr-Sender-Session".to_string(),
        message.sender_session_id.clone(),
    );
    headers.insert("Deckr-Recipient".to_string(), recipient_header(message));
    if let Some(value) = &message.recipient_session_id {
        headers.insert("Deckr-Recipient-Session".to_string(), value.clone());
    }
    if let Some(value) = &message.in_reply_to {
        headers.insert("Deckr-In-Reply-To".to_string(), value.clone());
    }
    headers
}

pub fn payload_json_bytes(message: &DeckrMessage) -> Result<Vec<u8>, serde_json::Error> {
    serde_json::to_vec(message)
}

pub fn state_payload_json_bytes<T: Serialize + ?Sized>(
    value: &T,
) -> Result<Vec<u8>, serde_json::Error> {
    serde_json::to_vec(value)
}

pub fn validate_subject_hint(
    subject: &str,
    message: &DeckrMessage,
) -> Result<(), NatsBindingError> {
    let prefix = format!("{LANE_PREFIX}.");
    if !subject.starts_with(&prefix) {
        return Ok(());
    }
    if subject_for(message) == subject {
        Ok(())
    } else {
        Err(NatsBindingError::SubjectMismatch)
    }
}

pub fn validate_headers(
    headers: Option<&BTreeMap<String, String>>,
    message: &DeckrMessage,
) -> Result<(), NatsBindingError> {
    let Some(headers) = headers else {
        return Ok(());
    };
    let expected = headers_for(message);
    for (key, expected_value) in expected {
        if headers
            .get(&key)
            .is_some_and(|actual| actual != &expected_value)
        {
            return Err(NatsBindingError::HeaderMismatch(key));
        }
    }
    Ok(())
}

File: libraries/typescript/src/nats.ts
import { compactJsonBytes, compactJsonString, contractJsonString } from "./json.js";
import {
  parseEndpointAddress,
  type DeckrMessage,
  type MessageTarget,
} from "./identity.js";
import { encodeKeyToken } from "./keys.js";

export const LANE_SUBJECT_PREFIX = "deckr.lane";
export const LANE_SUBJECT_TEMPLATE =
  "deckr.lane.{lane}.{senderFamily}.{senderEndpointToken}";
export const LANE_SUBSCRIBE_TEMPLATE = "deckr.lane.{lane}.>";
export const NATS_BINDING_SCHEMA_ID = "dev.deckr.binding.nats.v1";
export const NATS_BINDING_PATH = "bindings/nats.v1.json";
export const DECKR_NATS_HEADERS = [
  "Deckr-Message-Id",
  "Deckr-Message-Type",
  "Deckr-Sender",
  "Deckr-Sender-Session",
  "Deckr-Recipient",
  "Deckr-Recipient-Session",
  "Deckr-In-Reply-To",
] as const;
export const REQUIRED_DECKR_NATS_HEADERS = [
  "Deckr-Message-Id",
  "Deckr-Message-Type",
  "Deckr-Sender",
  "Deckr-Sender-Session",
  "Deckr-Recipient",
] as const;

export interface HeaderReader {
  get(name: string): string | null | undefined;
}

export function subjectFor(message: DeckrMessage): string {
  const sender = parseEndpointAddress(message.sender);
  if (sender === null) {
    throw new Error(`invalid sender endpoint: ${message.sender}`);
  }
  return [
    LANE_SUBJECT_PREFIX,
    encodeKeyToken(message.lane),
    encodeKeyToken(sender.family),
    encodeKeyToken(sender.endpointId),
  ].join(".");
}

export function subscribeSubjectForLane(lane: string): string {
  return `${LANE_SUBJECT_PREFIX}.${encodeKeyToken(lane)}.>`;
}

export function recipientHeader(message: DeckrMessage): string {
  return recipientHeaderValue(message.recipient);
}

export function headersFor(message: DeckrMessage): Record<string, string> {
  return {
    "Deckr-Message-Id": message.messageId,
    "Deckr-Message-Type": message.messageType,
    "Deckr-Sender": message.sender,
    "Deckr-Sender-Session": message.senderSessionId,
    "Deckr-Recipient": recipientHeader(message),
    ...(message.recipientSessionId === undefined
      ? {}
      : { "Deckr-Recipient-Session": message.recipientSessionId }),
    ...(message.inReplyTo === undefined
      ? {}
      : { "Deckr-In-Reply-To": message.inReplyTo }),
  };
}

export const deckrHeaderRecord = headersFor;
export const natsSubjectFor = subjectFor;

export function payloadJsonBytes(message: DeckrMessage): Uint8Array {
  return new TextEncoder().encode(payloadJsonString(message));
}

export function payloadJsonString(message: DeckrMessage): string {
  return contractJsonString(orderedMessage(message));
}

export function statePayloadJsonBytes(value: Record<string, unknown>): Uint8Array {
  return compactJsonBytes(value);
}

export function statePayloadJsonString(value: Record<string, unknown>): string {
  return compactJsonString(value);
}

export function validateHeaders(
  headers: HeaderReader | undefined,
  message: DeckrMessage,
): void {
  if (headers === undefined) {
    return;
  }
  for (const [name, expected] of Object.entries(headersFor(message))) {
    const actual = headers.get(name);
    if (actual !== undefined && actual !== null && actual !== expected) {
      throw new Error(`NATS header ${name} disagrees with Deckr envelope`);
    }
  }
}

export const validateHeaderHints = validateHeaders;

export function validateSubjectHint(subject: string, message: DeckrMessage): void {
  if (!subject.startsWith(`${LANE_SUBJECT_PREFIX}.`)) {
    return;
  }
  const actual = subject.split(".");
  const expected = subjectFor(message).split(".");
  if (actual.slice(0, expected.length).join(".") !== expected.join(".")) {
    throw new Error("NATS subject disagrees with Deckr envelope sender");
  }
}

function recipientHeaderValue(recipient: MessageTarget): string {
  if (recipient.targetType === "endpoint") {
    return recipient.endpoint;
  }
  return `broadcast:${recipient.scope}:${recipient.endpointFamily}`;
}

function orderedMessage(message: DeckrMessage): DeckrMessage {
  return {
    messageId: message.messageId,
    protocolVersion: message.protocolVersion,
    schemaVersion: message.schemaVersion,
    lane: message.lane,
    messageType: message.messageType,
    sender: message.sender,
    senderSessionId: message.senderSessionId,
    recipient: orderedRecipient(message.recipient),
    ...(message.recipientSessionId === undefined
      ? {}
      : { recipientSessionId: message.recipientSessionId }),
    subject: {
      kind: message.subject.kind,
      identifiers: message.subject.identifiers,
    },
    createdAt: message.createdAt,
    ...(message.expiresAt === undefined ? {} : { expiresAt: message.expiresAt }),
    ...(message.ttlMs === undefined ? {} : { ttlMs: message.ttlMs }),
    ...(message.inReplyTo === undefined ? {} : { inReplyTo: message.inReplyTo }),
    ...(message.causationId === undefined ? {} : { causationId: message.causationId }),
    ...(message.trace === undefined ? {} : { trace: message.trace }),
    body: message.body,
  };
}

function orderedRecipient(recipient: MessageTarget): MessageTarget {
  if (recipient.targetType === "endpoint") {
    return {
      targetType: "endpoint",
      endpoint: recipient.endpoint,
    };
  }
  return {
    targetType: "broadcast",
    scope: recipient.scope,
    endpointFamily: recipient.endpointFamily,
    ...(recipient.domain === undefined ? {} : { domain: recipient.domain }),
    ...(recipient.hopLimit === undefined ? {} : { hopLimit: recipient.hopLimit }),
  };
}
